### PR TITLE
OP-999 make jasper reports loadable from folder

### PR DESCRIPTION
**Pull request by nlazzer1**
See OP-999.

Changed folder structure for reports (rpt_base(hardcoded), rpt_stat (loadable folder) and rpt_extra(loadable folder).
The statistic module shows the reports contained in rpt_stat/ and rpt_user/ folders only.
The type of interval time (date_from-date_to or month) shown to the user is defined into the report by specific parameters.
The name of the report shown to the user is defined by the property named "jTitle" in the .properties file.

Check fails because need https://github.com/informatici/openhospital-core/pull/851.

### CODE REVIEW

File: rpt_stat/OH010_OutPatientReport_sv_SE.properties [hunk 1]
```diff
@@ -2,11 +2,7 @@
 
 Address = ADRESS
 
-Adult = Vuxen
-
-Age = ÅLDER
-
-Attendance = NÄRVARO
+Attendance = N\u00C3\u0084RVARO
```
Comment (mwithi) [inline rpt_stat/OH010_OutPatientReport_sv_SE.properties:5]: Please check files are not modified when moved (encoding UTF-8)

File: src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java [hunk 2]
```diff
@@ -260,20 +289,20 @@ private JPanel getJParameterSelectionPanel() {
 			jMonthComboBox.addItem(MessageBundle.getMessage("angal.stat.november"));
 			jMonthComboBox.addItem(MessageBundle.getMessage("angal.stat.december"));
 
-			jMonthComboBox.setSelectedIndex(month - 1);
+			jMonthComboBox.setSelectedIndex(month-1);
```
Comment (mwithi) [inline src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java:292]:
```suggestion
			jMonthComboBox.setSelectedIndex(month - 1);
```

File: src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java [hunk 3]
```diff
@@ -260,20 +289,20 @@ private JPanel getJParameterSelectionPanel() {
...
 
 			jYearLabel = new JLabel("        " + MessageBundle.getMessage("angal.stat.year"));
 			jYearComboBox = new JComboBox<>();
 
 			for (int i = 0; i < 20; i++) {
-				jYearComboBox.addItem(String.valueOf(year - i));
+				jYearComboBox.addItem((year - i) + "");
```
Comment (mwithi) [inline src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java:298]: Why this change?

File: src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java [hunk 4]
```diff
@@ -260,20 +289,20 @@ private JPanel getJParameterSelectionPanel() {
...
 			}
 			
-			jFromDateLabel = new JLabel(MessageBundle.getMessage("angal.common.datefrom.label"));
-			LocalDate defaultDate = LocalDate.now().minusMonths(8L);
+			jFromDateLabel = new JLabel(MessageBundle.getMessage("angal.stat.fromdate"));
+			LocalDate defaultDate = LocalDate.now().minusMonths(8);
```
Comment (mwithi) [inline src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java:302]: I think it is correct the original one

File: src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java [hunk 5]
```diff
@@ -349,56 +378,48 @@ private JButton getJCSVButton() {
...
+				new GenericReportMY(month, year, 
+								folderNameFileNameMap.get("rpt_stat").contains(reportNameFileMap.get(jRptComboBox.getSelectedItem().toString()).getName().replace(".jasper", "")) ? "rpt_stat" : "rpt_extra",
+								reportNameFileMap.get(jRptComboBox.getSelectedItem().toString()).getName().replace(".jasper", ""), 
+								jRptComboBox.getSelectedItem().toString(), toExcel);
 				if (GeneralData.XMPPMODULEENABLED) {
 					String user = (String) shareWith.getSelectedItem();
 					CommunicationFrame frame = (CommunicationFrame) CommunicationFrame.getFrame();
-					frame.sendMessage("011100100110010101110000011011110111001001110100 " + month + ' ' + year + ' ' + reportMatrix[rptIndex][FILENAME],
-							user, false);
+					frame.sendMessage("011100100110010101110000011011110111001001110100 " + month + " " + year +
```
Comment (mwithi) [inline src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java:415]: for single char strings we use char `' '`

File: src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java (hunk 4 above)
Comment (mwithi) [reply src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java:302]: @dbmalkovsky what about the LONG value?

File: src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java [hunk 6]
```diff
@@ -234,9 +219,53 @@ private JPanel getJParameterSelectionPanel() {
...
+                                .filter(t -> t.getName().endsWith(".jasper"))
+                                .collect(Collectors.toList());
+				jasperFilesInFolder.addAll(loadedFiles);
+				folderNameFileNameMap.put("rpt_extra", loadedFiles.stream().map(t -> t.getName().replace(".jasper", "")).collect(Collectors.toList()));
+			
+				reportNameFileMap = new HashMap<>();
+				for(File f: jasperFilesInFolder) {
+					try {
+						Properties props = ConfigurationProperties.loadPropertiesFile(
+										f.getName().replace(".jasper", "_"+(new Locale(GeneralData.LANGUAGE)).getLanguage()+".properties"), LOGGER);
```
Comment (mwithi) [inline src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java:245]:
```suggestion
										f.getName().replace(".jasper", "_" + (new Locale(GeneralData.LANGUAGE)).getLanguage() + ".properties"), LOGGER);
```

File: src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java [hunk 7]
```diff
@@ -234,9 +219,53 @@ private JPanel getJParameterSelectionPanel() {
...
+				loadedFiles = Files.walk(Paths.get("./rpt_extra"))
+                                .filter(Files::isRegularFile)
+                                .map(Path::toFile)
+                                .filter(t -> t.getName().endsWith(".jasper"))
+                                .collect(Collectors.toList());
+				jasperFilesInFolder.addAll(loadedFiles);
+				folderNameFileNameMap.put("rpt_extra", loadedFiles.stream().map(t -> t.getName().replace(".jasper", "")).collect(Collectors.toList()));
+			
+				reportNameFileMap = new HashMap<>();
+				for(File f: jasperFilesInFolder) {
```
Comment (mwithi) [inline src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java:242]:
```suggestion
				for (File f: jasperFilesInFolder) {
```

File: src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java [hunk 8]
```diff
@@ -234,9 +219,53 @@ private JPanel getJParameterSelectionPanel() {
...
+					try {
+						Properties props = ConfigurationProperties.loadPropertiesFile(
+										f.getName().replace(".jasper", "_"+(new Locale(GeneralData.LANGUAGE)).getLanguage()+".properties"), LOGGER);
+						
+						if(props != null && props.getProperty("jTitle") != null && !props.getProperty("jTitle").isEmpty()) {
```
Comment (mwithi) [inline src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java:247]:
```suggestion
						if (props != null && props.getProperty("jTitle") != null && !props.getProperty("jTitle").isEmpty()) {
```

File: src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java [hunk 9]
```diff
@@ -234,9 +219,53 @@ private JPanel getJParameterSelectionPanel() {
...
+							reportNameFileMap.put(props.getProperty("jTitle"), f);
+							jRptComboBox.addItem(props.getProperty("jTitle"));
+						}
+						else {
```
Comment (mwithi) [inline src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java:251]:
```suggestion
						} else {
```

File: src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java [hunk 10]
```diff
@@ -234,9 +219,53 @@ private JPanel getJParameterSelectionPanel() {
...
+							props = ConfigurationProperties.loadPropertiesFile(
+											f.getName().replace(".jasper", ".properties"), LOGGER);
+							
+							if(props != null && props.getProperty("jTitle") != null && !props.getProperty("jTitle").isEmpty()) {
+								reportNameFileMap.put(props.getProperty("jTitle"), f);
+								jRptComboBox.addItem(props.getProperty("jTitle"));
+							}
```
Comment (mwithi) [inline src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java:258]: looks like if jTitle is missing in all .properties, the report is not shown nor added to the list (differently as stated in the doc). Nevertheless, this could be good in order to force the presence of jTitle key.

Anyway, what we should be able to achieve is to show them also in alphabetical order

image

Comment (nlazzer1) [reply src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java:258]: If jTitle is missing in both localized and default .properties, now the report is added to the list with the name of the report itself and the report list is now shown in alphabetical order

File: src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java (hunk 9 above)
Comment (nlazzer1) [reply src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java:251]: I applied the OH code style

File: src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java (hunk 8 above)
Comment (nlazzer1) [reply src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java:247]: It seems that the rebase command did not update the file with the changes made by David. Now I have aligned the two files

File: src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java (hunk 9 above)
Comment (nlazzer1) [reply src/main/java/org/isf/stat/reportlauncher/gui/ReportLauncher.java:251]: Ok

**Review by mwithi — APPROVED** [review]
LGTM. Thanks a lot!
